### PR TITLE
Fix: Correct PW KPAR/BNDPAR occupations and distributed output

### PR DESCRIPTION
**Pull request by AsTonyshment**
## Summary

This PR fixes several PW KPAR/BNDPAR correctness issues. The changes correct fixed and explicitly supplied occupations under distributed k-point and band layouts, reconstruct complete band data before shared-file output, and fix the real-space grid collection used by charge-density `.cube` output.

The `kpar`, `bndpar`, and `ocp_set` parameter descriptions are also updated to match the current implementation.

## Problems addressed

- With `smearing_method=fixed` and `bndpar>1`, every band group treated its local band zero as global band zero. Multiple groups therefore occupied their first local bands repeatedly, producing incorrect electron counts, charge densities, band energies, and SCF results.
- `ocp_set` validated its input length using the pool-local number of k-points and indexed the input with a local-band stride. KPAR could reject a correctly sized input, while BNDPAR could silently assign occupations to the wrong bands.
- `out_chg` used source ranks derived from a larger process layout as ranks in `POOL_WORLD`. Under BPCG this could pass an invalid communicator rank to `MPI_Recv` after the SCF calculation had completed.
- `out_band` and eigenvalue/occupation output treated band-local matrices as if they contained all global bands. This could cause out-of-bounds access, omit bands owned by other band groups, duplicate k-point contributions, or allow multiple groups to write the same shared file.

## Validation

A standalone edge-case test was carried out with 21 calculations: 17 CPU cases and 4 GPU cases. PW CPU cases used 6 MPI ranks, 8 k-points, 7 bands, and BPCG. The following layouts were compared:

| Layout | Coverage |
|---|---|
| `kpar=1,bndpar=1` | Reference layout |
| `kpar=2,bndpar=1` | KPAR only |
| `kpar=1,bndpar=3` | Uneven band ownership, `3+2+2` |
| `kpar=2,bndpar=3` | Mixed KPAR/BNDPAR |
| `kpar=3,bndpar=2` | Uneven k-point pools, `3+3+2`, and uneven band ownership, `4+3` |

The edge-case test covered `nspin=1` and `nspin=2`, fixed occupations, k-point-dependent `ocp_set` values, PW CPU/GPU execution, and LCAO temporary `kpar=1/2/3` diagonalization. Every PW case checked the complete occupation table, total energy, band table, charge-density `.cube` values, and `.cube` integral.

All 21 cases completed successfully and reported no failures.

| Test group | Maximum energy difference / eV | Maximum band-table difference | Maximum Cube point difference | Maximum Cube integral difference |
|---|---:|---:|---:|---:|
| PW CPU | `7.190692e-12` | `1.534826e-10` | `2.299966e-13` | `1.199041e-14` |
| LCAO CPU | `8.526513e-14` | `1.163514e-13` | `1.000589e-14` | `2.042810e-14` |
| PW GPU | `0.000000e+00` | `3.730349e-14` | `1.000589e-14` | `5.329071e-15` |

### CODE REVIEW

**Comment by mohanchen** [comment]
Could you add some unit tests or integrate tests?

**Comment by AsTonyshment** [comment]
> Could you add some unit tests or integrate tests?

Added. I also found that `get_pchg_pw` and `get_wf_pw` may obtain incorrect results under BNDPAR (I only considered KPAR in the original implementation), and this is fixed as well. Relevant tests are updated.

File: source/source_estate/elecstate_tools.cpp [hunk 1]
```diff
@@ -170,11 +189,28 @@ namespace elecstate
                                     "total number of occupations is wrong , please check ocp_set");
         }
 
+        const int band_offset = get_band_offset(wg.nc);
```
[the remark this answers is not in the document]
Comment (AsTonyshment) [reply source/source_estate/elecstate_tools.cpp]: This finding is correct. Non-BPCG SDFT keeps the complete deterministic band set in every band group, so applying an unconditional `BP_WORLD` prefix sum gives nonzero groups an invalid offset even though their matrices are replicated.

I fixed the layout detection used by occupation assignment. The helper now gathers all local band counts on `BP_WORLD`, returns offset zero when every group contains `global_nbands`, and otherwise validates that the complementary shards sum to `global_nbands` before calculating the prefix offset. The same rule is now used by both `fixed_weights()` for `ocp_set` and the integer-weight path for `smearing_method=fixed`. `calculate_weights()` receives `global_nbands` explicitly.

I also added the 4-rank `MODULE_ESTATE_band_parallel_weights_test_4np` regression. It checks both replicated `5+5` storage and uneven distributed `3+2` storage for explicit fixed occupations and integer weights. The new MPI test and the existing `MODULE_ESTATE_elecstate_base` test both pass.

File: source/source_base/parallel_grid.cpp [hunk 2]
```diff
@@ -422,60 +422,81 @@ void Parallel_Grid::zpiece_to_stogroup(double* zpiece, const int& iz, double* rh
...
-            }
-        }
+    int pool_size = 0;
+    MPI_Comm_size(POOL_WORLD, &pool_size);
+
+    // Derive slab ownership from the communicator used below. Precomputed
+    // source ranks may belong to a larger BPCG layout and are not necessarily
+    // valid ranks in the current POOL_WORLD.
+    std::vector<int> local_z_counts(pool_size);
+    MPI_Allgather(&this->nczp, 1, MPI_INT, local_z_counts.data(), 1, MPI_INT, POOL_WORLD);
```
[the remark this answers is not in the document]
Comment (AsTonyshment) [reply source/source_base/parallel_grid.cpp:446]: I checked the current implementation and do not believe the reported mismatch exists in `Parallel_Grid::reduce()`.

Although `Parallel_Grid::init()` retains the older `numz` and `startz` tables, the current `reduce()` does not index either table. It gathers each rank's actual `nczp` on the same `POOL_WORLD` communicator used by `MPI_Gatherv`, then derives all receive counts, rank-block displacements, and global z starts exclusively from that communicator-local vector. The `pw_rhod` basis supplying `nczp` and `nrxx` is itself initialized with `NPROC_IN_POOL`, `RANK_IN_POOL`, and `POOL_WORLD`, so the gathered distribution matches the collective participants.

I therefore did not change this function for the review comment.

File: tests/01_PW/087_PW_get_pchg_kpar_bndpar/pchgi4s1.cube.ref
Comment (mohanchen) [inline tests/01_PW/087_PW_get_pchg_kpar_bndpar/pchgi4s1.cube.ref:1]: the file seems big, could you reduce the size of the file? (lower ecut)

Comment (AsTonyshment) [reply tests/01_PW/087_PW_get_pchg_kpar_bndpar/pchgi4s1.cube.ref:1]: Thanks for the suggestion. Since `ecutwfc = 20` Ry is already relatively low, I reduced the lattice constant instead. This reduces the FFT grid and shrinks `pchgi4s1.cube.ref` from 2312 to 980 lines.

**Review by mohanchen — APPROVED** [review]
LGTM
